integrations/langchain4j/langchain4j/src/main/java/io/helidon/integrations/langchain4j/MetricsChatModelListener.java

```diff
@@ -0,0 +1,181 @@
+
+/**
+ * Creates metrics that follow the
+ * <a href="https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics/">Semantic Conventions for GenAI Metrics</a>.
+ */
+@Service.Singleton
+@Service.Named("MetricsChatModelListener")
+public class MetricsChatModelListener implements ChatModelListener {
+
+    private static final String GEN_AI_CLIENT_OPERATION_START_TIME = "GEN_AI_CLIENT_OPERATION_START_TIME";
+    private static final String GEN_AI_CLIENT_TOKEN_USAGE_METRICS_NAME = "gen_ai.client.token.usage";
+    private static final String GEN_AI_CLIENT_OPERATION_DURATION_METRICS_NAME = "gen_ai.client.operation.duration";
+
+    private final MeterRegistry meterRegistry;
+    private final DistributionStatisticsConfig.Builder clientTokenUsageStatisticsConfigBuilder;
+    private final DistributionStatisticsConfig.Builder clientOperationDurationStatisticsConfigBuilder;
+    private final Map<String, DistributionSummary> responseInputTokenUsageSByModelName = new ConcurrentHashMap<>();
+    private final Map<String, DistributionSummary> responseOutputTokenUsageSByModelName = new ConcurrentHashMap<>();
+    private final Map<String, DistributionSummary> responseOperationDurationByModelName = new ConcurrentHashMap<>();
+    private final Map<String, DistributionSummary> errorOperationDurationByModelName = new ConcurrentHashMap<>();
+
+    /**
+     * Constructs a {@code MetricsChatModelListener} instance.
+     */
+    public MetricsChatModelListener() {
+        this.meterRegistry = Metrics.globalRegistry();
+
+        // Limits set based on https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics/#metric-gen_aiclienttokenusage.
+        this.clientTokenUsageStatisticsConfigBuilder = DistributionStatisticsConfig.builder()
+                .buckets(1, 4, 16, 64, 256, 1024, 4096, 16384, 65536, 262144, 1048576, 4194304, 16777216, 67108864);
+
+        // Limits set based on https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-metrics/#metric-gen_aiclientoperationduration.
+        this.clientOperationDurationStatisticsConfigBuilder =
+                DistributionStatisticsConfig.builder()
+                        .buckets(0.01, 0.02, 0.04, 0.08, 0.16, 0.32, 0.64, 1.28, 2.56, 5.12, 10.24, 20.48, 40.96, 81.92);
+    }
+
+    @Override
+    public void onRequest(ChatModelRequestContext chatModelRequestContext) {
+        chatModelRequestContext.attributes().put(GEN_AI_CLIENT_OPERATION_START_TIME, System.nanoTime());
+    }
+
+    @Override
+    public void onResponse(ChatModelResponseContext chatModelResponseContext) {
+        final long endTime = System.nanoTime();
+        final long startTime = (Long) chatModelResponseContext.attributes().get(GEN_AI_CLIENT_OPERATION_START_TIME);
+
+        final ChatRequest chatRequest = chatModelResponseContext.chatRequest();
+        final ChatResponse chatResponse = chatModelResponseContext.chatResponse();
+
+        responseInputTokenUsageSByModelName.computeIfAbsent(
```

Comment:
In each case, save the return value from the `computeIfAbsent` call in a variable because it returns either the previously-stored value in the map or the new, just-stored one. Then invoke `record` on that variable, avoiding the need to look up the metric again in the next statement.

Otherwise this looks very good. It should perform much better.